test/model/internal_link_test.dart

```diff
@@ -0,0 +1,486 @@
+    //     ('/#narrow/is/mentioned', ...),
+    //   ];
+    //   checkExpectedNarrows(testCases, store: store);
+    // });
+
+    test('unexpected link shapes are rejected', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('/#narrow/stream/stream/topic/topic/near/', null),
+        // ('/#narrow/is/men', null),
+        // ('/#narrow/is/men/stream', null),
+        ('/#narrow/are/men/stream', null),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+  });
+
+  group('decodeHashComponent', () {
+    test('correctly decode MediaWiki-style dot-encoded strings', () {
+      check(decodeHashComponent('some_text')).equals('some_text');
+      check(decodeHashComponent('some.20text')).equals('some text');
+      check(decodeHashComponent('some.2Etext')).equals('some.text');
+
+      check(decodeHashComponent('na.C3.AFvet.C3.A9')).equals('naïveté');
+      check(decodeHashComponent('.C2.AF.5C_(.E3.83.84)_.2F.C2.AF')).equals('¯\\_(ツ)_/¯');
+
+      // malformed dot-encoding
+      check(decodeHashComponent('some.text')).isNull();
+      check(decodeHashComponent('some.2gtext')).isNull();
+      check(decodeHashComponent('some.arbitrary_text')).isNull();
+
+      // malformed UTF-8
+      check(decodeHashComponent('.88.99.AA.BB')).isNull();
+    });
+
+    test('parse correctly in stream and topic operands', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'some_stream'),
+        eg.stream(streamId: 2, name: 'some stream'),
+        eg.stream(streamId: 3, name: 'some.stream'),
+      ];
+      final testCases = [
+        ('/#narrow/stream/some_stream',
+          const StreamNarrow(1)),
+        ('/#narrow/stream/some.20stream',
+          const StreamNarrow(2)),
+        ('/#narrow/stream/some.2Estream',
+          const StreamNarrow(3)),
+        ('/#narrow/stream/some_stream/topic/some_topic',
+          const TopicNarrow(1, 'some_topic')),
+        ('/#narrow/stream/some_stream/topic/some.20topic',
+          const TopicNarrow(1, 'some topic')),
+        ('/#narrow/stream/some_stream/topic/some.2Etopic',
+          const TopicNarrow(1, 'some.topic')),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+  });
+
+  group('Stream link parsing edge cases', () {
+    final streams = [
+      eg.stream(streamId: 1, name: "general"),
+    ];
+
+    test('basic', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/1-general', const StreamNarrow(1)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
+
+    test('if stream not found, use stream ID anyway', () {
+      final store = setupStore(realmUrl: realmUrl, streams: streams);
+      final testCases = [
+        ('#narrow/stream/123-topic', const StreamNarrow(123)),
+      ];
+      checkExpectedNarrows(testCases, store: store);
+    });
```

Comment:
Compare these lines to the corresponding lines in the zulip-mobile tests:

```js
    test('basic', () => {
      expectStream(`${streamGeneral.stream_id}-general`, [streamGeneral], streamGeneral);
    });

    test('if stream not found, use stream ID anyway', () => {
      expectStream(`${streamGeneral.stream_id}-general`, [], streamGeneral);
    });
```

In the latter, the two lines with the details of the tests are separated by 1 line of text and 2 nearly-blank lines. In this version, they're separated by 7 lines, of which 4 have text on them. That makes it a lot noisier for the eye to compare the two and see what the contrasts are. Similarly for the next few cases below.

For this sort of thing, it can really help a lot to make a small local helper function. That's what `expectStream` in the corresponding zulip-mobile code is doing.